### PR TITLE
optim/best-degree: set the optimal gate degree

### PR DESCRIPTION
**Pull request by naure**
The gate degree of the EVM circuit is currently 7. This is set by the gate "Constrain state machine transitions". However, the constraint builder limits other gates to 6 instead of 7. The parameters and the comment seem outdated.

Meanwhile, 7 is a poor trade-off. The next threshold is **9, which is chosen in this PR**.

The effective degree of gadgets goes from 2 to 5. This will reduce the usage of cells for stored expressions. This, in turn, allows someone to simply configure fewer "Storage" columns, saving verifier time.

**Stats!**

- The gate complexity is down 17% (from 10623 to 8838). This is measured as the count of multiplications, after memoizing. This is the bottleneck in prover time.

Observing the stored expressions of some individual states:

- CALLDATALOAD is no longer the largest instruction. It uses ~2X fewer stored expressions, moving to its next bottleneck at a lower height.
- PUSH -68%. MEMORY -63%. SIGNEXTEND -51%. MUL -20%. ADD -12%. BITWISE -10%.

Data using #863 -- [Before](https://github.com/privacy-scaling-explorations/zkevm-circuits/files/9908474/opcode_metrics3_degree6.txt) -- [After](https://github.com/privacy-scaling-explorations/zkevm-circuits/files/9908475/opcode_metrics3_degree9.txt).

### CODE REVIEW

**Comment by ed255** [comment]
Thanks for the analysis! The EVM circuit degree was recently increased in this PR https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/718 where I explain a bit the rationale behind going to a non-optimal gate degree: the plan is to reduce 1 degree once we remove the blinding factors by removing the `q_usable` selector from the expressions.

Nevertheless I thought that the current max gate degree in EVM was 6: 5 + 1.  If it's 7, even if we reduce 1 degree by removing `q_usable` we're still at extended k = 3.  I've checked where the 7 comes from and found the reason (I introduced that extra +1 as well):
https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/b87a9d2a08d88cd9d04ba64a5e7ca3c7f6e34471/zkevm-circuits/src/evm_circuit/execution.rs#L687-L691

The degrees are:
- q_usable: 1
- q_step: 1
- (1 - q_step_last): 1
- execution_state_selector: 2
- poly: 2

I guess we could find a creative way bring the degree down (like adding an extra column to hold `q_step * (1 - q_step_last)`.

In summary: the current degree was supposed to be temporary and we were aiming for degree 5 in the mid-term.  We may decide to go to degree 9 but we need to consider the proving performance implications.

**Comment by naure** [comment]
Couldn’t instead q_step be constrained by q_usable?

    if q_usable == 0 then q_step = 0

And there q_usable won’t need to multiply all gates because q_step is already doing it.

**Comment by ed255** [comment]
> Couldn’t instead q_step be constrained by q_usable?
> 
> ```
> if q_usable == 0 then q_step = 0
> ```
> 
> Then q_usable won’t need to multiply all gates because q_step is already doing it.

Yes!  That would work and we'd remove `q_usable` with it.  Now we only need to remove one more factor to get the degree at 5.  If we can turn `q_step * (1 - q_step_last)` into an expression of degree 1 that would do it.  Currently we need a way to distinguish all the steps from the last one because all steps must enforce a correct transition to the next one, except for the last (which doesn't have a next step).  Maybe we could have an extra EndBlock after the last EndBlock with `q_usable = 0, q_step = 0` so that the last step can have transition constraints to the next fulfilled?  Then we'd remove `1 - q_step_last` from the expression, and enable the EndBlock transition constrains in all steps.

Having said that, I'm afraid at some point we may need to have constraints that not "stateless" enforced in the last EndBlock, like sending a reward to the miner (which increases the rwc), although... we could have multiple EndBlocks doing that as long as we don't force an rwc increase for the next step (they would just verify the same state change over and over, using the same rwc values).

@han0110 what do you think about this?  Is the idea clear?

**Comment by han0110** [comment]
> Couldn’t instead q_step be constrained by q_usable?
> 
> ```
> if q_usable == 0 then q_step = 0
> ```
> 
> Then q_usable won’t need to multiply all gates because q_step is already doing it.

This should be possible and desirable when we remove the blinding factor.

> Adding an extra column to hold `q_step * (1 - q_step_last)`.

This seems could be temporary solution for now.

**Comment by ChihChengLiang** [comment]
Hi, @naure @ed255 @han0110 what's the status of this PR? Does anything need more clarification?

**Review by ed255 — APPROVED** [review]
First of all, sorry for taking so long to get back to this!

Currently we have many little tasks here and there to have a more complete zkEVM implementation, so we're not focusing too much on these kind of optimizations.  Specially what I discussed in https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/869#issuecomment-1303827543  So I think we're not in a hurry to make the changes I wrote in my comment; and we can get back to them once the development of missing parts slows down a bit.

The changes in your PR are very simple and I realized that even if we want to reduce the constraint degree to 5 in the future, we will live with >5 for a while until we focus on that task; so it makes total sense to let the `ConstraintBuilder` work with degree 9 for now to get the performance benefit!

So LGTM!

**Comment by ed255** [comment]
@icemelon could you do a second review to this PR? It's a small change that gives some performance benefits (and I think it's useful for the time until we bring constraint the degree down to 5)

**Comment by CPerezz** [comment]
I did not notice that this was assigned to me apologies!!! Let me know if you still want me to look at this!

File: zkevm-circuits/src/evm_circuit/util/constraint_builder.rs [hunk 1]
```diff
@@ -23,11 +23,11 @@ use halo2_proofs::{
...
-// It aims to cap `extended_k` to 2, which allows constraint degree to 2^2+1,
-// but each ExecutionGadget has implicit selector degree 3, so here it only
-// allows 2^2+1-3 = 2.
-const MAX_DEGREE: usize = 5;
-const IMPLICIT_DEGREE: usize = 3;
+// It aims for `extended_k = k + 3`, which allows constraint degree to 2^3+1=9,
+// but each ExecutionGadget has implicit selector degree 4, so here it only
+// allows 9-4=5.
+const MAX_DEGREE: usize = 9; // 2^3 + 1
+const IMPLICIT_DEGREE: usize = 4; // q_usable + q_step + 2 state selectors
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/util/constraint_builder.rs:30]: I think the current implicit degree is 5, not 4. So we should correct it here to 5, right?

File: zkevm-circuits/src/evm_circuit/util/constraint_builder.rs [hunk 2]
```diff
@@ -23,11 +23,11 @@ use halo2_proofs::{
 use super::{rlc, CachedRegion, CellType, StoredExpression};
 
 // Max degree allowed in all expressions passing through the ConstraintBuilder.
-// It aims to cap `extended_k` to 2, which allows constraint degree to 2^2+1,
-// but each ExecutionGadget has implicit selector degree 3, so here it only
-// allows 2^2+1-3 = 2.
-const MAX_DEGREE: usize = 5;
-const IMPLICIT_DEGREE: usize = 3;
+// It aims for `extended_k = k + 3`, which allows constraint degree to 2^3+1=9,
```
Comment (ed255) [inline zkevm-circuits/src/evm_circuit/util/constraint_builder.rs:26]: Shouldn't it be `extended_k = 3`?

File: zkevm-circuits/src/evm_circuit/util/constraint_builder.rs (hunk 1 above)
Comment (ed255) [reply zkevm-circuits/src/evm_circuit/util/constraint_builder.rs:30]: Ah I think you're right.  It should be
```rust
const IMPLICIT_DEGREE: usize = 5; // q_usable + q_step + q_step_last + 2 state selectors
```

**Comment by aguzmant103** [comment]
@naure just doing a ping as I think there are some open comments to solve before getting this to merge

**Comment by aguzmant103** [comment]
Applying the "won'tfix" label now because this optimization is not required now, but we can reconsider it once this optimization becomes more critical.

**Comment by leolara** [comment]
@aguzmant103 I close it then
